### ISSUE TITLE
Non-Deterministic OPTKING Steps with Dihedral 0

### ISSUE THREAD

**Issue by JonathonMisiewicz** [psi4/psi4#860]
Under certain conditions, OPTKING's steps are non-deterministic. A dihedral of 0 seems necessary, but it is not clear to me if this is sufficient.

I've only found this in the IRC code, though I haven't checked for this in the non-IRC code. The below three files use the exact same input file but all have different steps. (This is on my own ircfix branch. I had conv_check return false on the first iteration as described [here](https://github.com/psi4/psi4/issues/859) and hardcoded the filename for my external hessian, to speed up my tests. The hessian itself was generated at SCF with this geometry.)

[output.txt](https://github.com/psi4/psi4/files/1472907/output.txt)
[output1.txt](https://github.com/psi4/psi4/files/1472908/output1.txt)
[output2.txt](https://github.com/psi4/psi4/files/1472909/output2.txt)

The below cases do not show the non-deterministic behavior. The only change in input file from the previous runs to this run was that the dihedral of 0 has been changed to a dihedral of 2.

[output.txt](https://github.com/psi4/psi4/files/1472911/output.txt)
[output1.txt](https://github.com/psi4/psi4/files/1472912/output1.txt)

**Comment by psi-rking** [comment]
I don't know how you are reading in the Hessian here, since I don't know how to automatically read files like "hooh.21221.hess" now that we have pid's in the name.   (BTW, it appears the current psi4 test cases actually use an empirical Hessian.)

These jobs are failing because the "Point 1" on the IRC is being judged 'converged' immediately before a gradient is even run.  (The logic in the code is messy, since the optimizer doesn't actually call the gradient.)  So optking needs to do displacements from the TS (where the gradient is zero) but return a not-converged signal to the python optimize.  This usually works.  I tried the following:

```
set {
  g_convergence gau_verytight
  basis dzp
  opt_type                   irc
  cart_hess_read true
  hessian_write true
}

hessian('scf', dertype=1)
h2o2.reset_point_group('c2')
energy = optimize('scf')

```
and everything worked fine.  The problem in my testing is not indeterminism.  There is something about the calculation setup that is causing optking not to compute the gradient after the first irc_step().  Perhaps @avcopan could figure out what is going on?

**Comment by JonathonMisiewicz** [comment]
I'm already aware of the calculation setup not computing the gradient after the first irc_step(). I describe the mechanism for this in detail in #859. I played around with some fixes, but those did not work. Given that OPTKING should be brought Python side in the not-too-distant future, further investigating was much less of a priority for me.

I could swear I determined that this bug was at least partly separate from that bug, but I'd need to look over my files to be sure.

**Comment by psi-rking** [comment]
These could be two separate issues.  However, the second step of the IRC is in the direction of the gradient.  If optking thinks Point 1 is converged on the path, when it hasn't even computed a gradient yet, then the 'current' gradient still from the TS will be close to zero and might account for apparent randomness.

**Comment by avcopan** [comment]
I can confirm that your input, @psi-rking, seems to work fine.  I don't see any indeterminism either.  When I turn symmetry off entirely, though, the IRC does quit early and I'm not sure why.  I poked around the code a bit -- debugging this is a real headache, since I wrote it 6 years ago and the logic is pretty convoluted.

**Comment by psi-rking** [comment]
If going to C1 makes a difference, then a couple of likely reasons are that
1) we are symmetrizing the step if C2 is included; or 2) it's surprisingly
easy to get the wrong orbital occupations for hydrogen peroxide.  It will
be less convoluted in the new python version (which still needs
debugged...).

**Comment by JonathonMisiewicz** [comment]
I ran some tests with a hotfix of #859, and while there was still some variation, it was confined to the last decimal place, even after nine points on the path. So yes, this is a bug closely tied to #859 after all.

Note to Self:

Said hotfix was inserting the following line of code under the "Test for convergence" section of opt_data.cc:
`if (Opt_params.opt_type == OPT_PARAMS::IRC && g_iteration() == 1) return false;`
I recall some rare cases where the second iteration would be taken as the stationary point. I'm having a hard time reproducing this, but checking if the Hessian has an imaginary eigenvalue would fix it. I do not have the time at the moment to figure out how I would actually check that, in the C++ code.

This explains nothing about the symmetry problem, but I would be completely unsurprised if there was some symmetrizing step somewhere in this logic.

**Comment by JonathonMisiewicz** [comment]
@psi-rking @avcopan Some more thorough testing of the #859 hotfix above allowed me to reproduce non-deterministic behavior where the IRC may or may not converge after the second step. Output files below.

[output1.txt](https://github.com/psi4/psi4/files/1533781/output1.txt)
[output2.txt](https://github.com/psi4/psi4/files/1533782/output2.txt)

I want #859 fixed before trying to fix this issue, because that kills my test cases before they encounter this bug. I see two options for the #859 fix.

1. Forbid the IRC from ending if the hessian has a negative eigenvalue, as we cannot be at a minimum. This fixes both issues, but requires that Optking's hessian agree with the true hessian about whether there is a negative eigenvalue. Is this a reliable assumption? And what if Optking gets "stuck" at a very flat initial transition state? Do we want it to continue trying to get out until it exhausts its number of optimizations?
2. Keep the condition I tried to generate these output files, where we only forbid exit on iteration 1. It is up to the user to check that the IRC does not get stuck at the initial TS later and change the step size if it does get stuck.

In either case, fixing #859 will not fix this bug, so I'll have to investigate the source of this non-determinism next.

**Comment by JonathonMisiewicz** [comment]
@psi-rking 

I have the IRC mostly fixed, but the final step is difficult to control. For comparison, have the energies that result from the exact same job being run six times:

```@IRC ----------------------------------------------
@IRC            ****      IRC Report      ****
@IRC ----------------------------------------------
@IRC  Step    Energy              Change in Energy 
@IRC ----------------------------------------------
@IRC    0  -148.750562902668   -148.750562902668
@IRC    1  -148.750914020851     -0.000351118183
@IRC    2  -148.751921813112     -0.001007792261
@IRC    3  -148.753458618899     -0.001536805787
@IRC    4  -148.755338547024     -0.001879928125
@IRC    5  -148.757353153707     -0.002014606684
@IRC    6  -148.759306859301     -0.001953705594
@IRC    7  -148.761045995024     -0.001739135723
@IRC    8  -148.762474848946     -0.001428853922
@IRC    9  -148.763557315982     -0.001082467035
@IRC   10  -148.764306952597     -0.000749636616
@IRC   11  -148.764771570120     -0.000464617522
@IRC   12  -148.765017199697     -0.000245629578
@IRC   13  -148.765113810944     -0.000096611247
@IRC   14  -148.765124130978     -0.000010320034
@IRC   15  -148.764729734758      0.000394396220
@IRC ----------------------------------------------

@IRC ----------------------------------------------
@IRC            ****      IRC Report      ****
@IRC ----------------------------------------------
@IRC  Step    Energy              Change in Energy 
@IRC ----------------------------------------------
@IRC    0  -148.750562902668   -148.750562902668
@IRC    1  -148.750914020605     -0.000351117937
@IRC    2  -148.751921813098     -0.001007792493
@IRC    3  -148.753458618748     -0.001536805651
@IRC    4  -148.755338547153     -0.001879928405
@IRC    5  -148.757353153742     -0.002014606589
@IRC    6  -148.759306859354     -0.001953705612
@IRC    7  -148.761045994942     -0.001739135587
@IRC    8  -148.762474848816     -0.001428853874
@IRC    9  -148.763557315945     -0.001082467129
@IRC   10  -148.764306952816     -0.000749636871
@IRC   11  -148.764771569962     -0.000464617146
@IRC   12  -148.765017199740     -0.000245629777
@IRC   13  -148.765113810834     -0.000096611095
@IRC   14  -148.765124130995     -0.000010320161
@IRC   15  -148.764489218884      0.000634912111
@IRC ----------------------------------------------

@IRC ----------------------------------------------
@IRC            ****      IRC Report      ****
@IRC ----------------------------------------------
@IRC  Step    Energy              Change in Energy 
@IRC ----------------------------------------------
@IRC    0  -148.750562902668   -148.750562902668
@IRC    1  -148.750914020681     -0.000351118013
@IRC    2  -148.751921813016     -0.001007792335
@IRC    3  -148.753458618767     -0.001536805751
@IRC    4  -148.755338546974     -0.001879928208
@IRC    5  -148.757353153593     -0.002014606618
@IRC    6  -148.759306859284     -0.001953705691
@IRC    7  -148.761045994957     -0.001739135673
@IRC    8  -148.762474848905     -0.001428853949
@IRC    9  -148.763557315936     -0.001082467031
@IRC   10  -148.764306952814     -0.000749636878
@IRC   11  -148.764771570110     -0.000464617296
@IRC   12  -148.765017199687     -0.000245629578
@IRC   13  -148.765113811009     -0.000096611321
@IRC   14  -148.765124130941     -0.000010319932
@IRC   15  -148.765124130672      0.000000000269
@IRC ----------------------------------------------

@IRC ----------------------------------------------
@IRC            ****      IRC Report      ****
@IRC ----------------------------------------------
@IRC  Step    Energy              Change in Energy 
@IRC ----------------------------------------------
@IRC    0  -148.750562902668   -148.750562902668
@IRC    1  -148.750914020798     -0.000351118130
@IRC    2  -148.751921813074     -0.001007792276
@IRC    3  -148.753458618742     -0.001536805668
@IRC    4  -148.755338547032     -0.001879928290
@IRC    5  -148.757353153648     -0.002014606616
@IRC    6  -148.759306859381     -0.001953705733
@IRC    7  -148.761045995000     -0.001739135620
@IRC    8  -148.762474848907     -0.001428853907
@IRC    9  -148.763557315941     -0.001082467034
@IRC   10  -148.764306952859     -0.000749636918
@IRC   11  -148.764771569930     -0.000464617071
@IRC   12  -148.765017199626     -0.000245629697
@IRC   13  -148.765113810904     -0.000096611278
@IRC   14  -148.765124131311     -0.000010320407
@IRC   15  -148.765124131554     -0.000000000244
@IRC ----------------------------------------------

@IRC ----------------------------------------------
@IRC            ****      IRC Report      ****
@IRC ----------------------------------------------
@IRC  Step    Energy              Change in Energy 
@IRC ----------------------------------------------
@IRC    0  -148.750562902668   -148.750562902668
@IRC    1  -148.750914020614     -0.000351117946
@IRC    2  -148.751921813072     -0.001007792458
@IRC    3  -148.753458618806     -0.001536805734
@IRC    4  -148.755338546954     -0.001879928149
@IRC    5  -148.757353153393     -0.002014606439
@IRC    6  -148.759306859244     -0.001953705851
@IRC    7  -148.761045994807     -0.001739135563
@IRC    8  -148.762474848960     -0.001428854153
@IRC    9  -148.763557315823     -0.001082466863
@IRC   10  -148.764306952777     -0.000749636954
@IRC   11  -148.764771569804     -0.000464617027
@IRC   12  -148.765017199652     -0.000245629848
@IRC   13  -148.765113811017     -0.000096611365
@IRC   14  -148.765113820352     -0.000000009335
@IRC ----------------------------------------------

@IRC ----------------------------------------------
@IRC            ****      IRC Report      ****
@IRC ----------------------------------------------
@IRC  Step    Energy              Change in Energy 
@IRC ----------------------------------------------
@IRC    0  -148.750562902668   -148.750562902668
@IRC    1  -148.750914020808     -0.000351118140
@IRC    2  -148.751921813104     -0.001007792295
@IRC    3  -148.753458618713     -0.001536805609
@IRC    4  -148.755338547051     -0.001879928339
@IRC    5  -148.757353153581     -0.002014606530
@IRC    6  -148.759306859192     -0.001953705611
@IRC    7  -148.761045995024     -0.001739135832
@IRC    8  -148.762474848811     -0.001428853786
@IRC    9  -148.763557315832     -0.001082467021
@IRC   10  -148.764306952658     -0.000749636826
@IRC   11  -148.764771570088     -0.000464617430
@IRC   12  -148.765017199695     -0.000245629606
@IRC   13  -148.765113810604     -0.000096610909
@IRC   14  -148.765124130954     -0.000010320350
@IRC   15  -148.764489158578      0.000634972376
@IRC ----------------------------------------------
```

**I'm inclined to submit what I have now and marked these bugs as closed.** Opinions? There's still some non-determinism, but disagreement about the last step means the IRC is at least working now.

**Comment by avcopan** [comment]
I'm okay with that if the current tests pass.

Ideally there should be a test-case attached sooner rather than later.  I think a simple-but-intuitive way to do this would be to add a keyword named something like `irc_n_steps` which tells OPTKING to `return OptReturnSuccess` after a given number of converged IRC points.  Then you could do something like `set irc_n_steps 10` in your test case to avoid passing through the minimum, after which the steps may become unpredictable.  I could help with that.

**Comment by JonathonMisiewicz** [comment]
This issue was fixed by #882. While there is still some non-determinism when the optimizer gets to a minimum, it does not have a common cause with the behavior reported here. It may also be fixed when optking is brought Python-side.

For reference, the output I generated demonstrating non-determinism at the mimum was using the test case created in #882, without the early exit condition. If this issue needs to be raised again, it would be a separate issue.
